### ISSUE TITLE
Please add Software UART to ESP32

### ISSUE THREAD

**Issue by it-guy92** [esphome/feature-requests#1712]
**Describe the problem you have/What new integration you would like**

**Please describe your use case for this integration and alternatives you've tried:**

**Additional context**

**Comment by ShunichiSan** [comment]
I've encountered the same problem. The doc says "When the hardware UARTs are all occupied, ESPHome will fall back to a software implementation that may not be accurate at higher baud rates." 
I'm migrating  a project from an Arduino nano which easily handled 4 simultaneous DFplayers using softwareserial (each requires its own serial connection). However with ESPHome on an ESP32 only two will work regardless of my pin selection. The third UART is presumably being used by the logger.  Each DFPlayer works if I initiate them alone, but enabling more than two only results in the first two working. It would be great if the software UARTs were enabled, or at least update the docs.

**Comment by wawibu** [comment]
Glad to found this as I run into the same issue. I want to use an ESP32 with more than 3 UART and I also understood the documentation in the way that EspHome would fall back to sw uart. 
Using the hw uart already for GPS, MPPT charger and SDM120. Now I want to add the Nextion Display to it, but didn't get it working as it didn't fall back to sw uart if all hw uart already occupied.
Is there any option to force to sw uart?

**Comment by lwqcz** [comment]
I found this repo and it seems to be working ... https://github.com/KinDR007/VictronMPPT-ESPHOME

**Comment by rnizametdinov** [comment]
> I found this repo and it seems to be working ... https://github.com/KinDR007/VictronMPPT-ESPHOME

i think it use HW serial on ESP32. 
I dont found any custom softserial code in this repo.

**Comment by simbaja** [comment]
Ran into this issue as well.  ESP32-S2 only has two UARTs apparently, so it's even more problematic.  It caused me a bit of confusion, since the documentation says it'll fallback to software, but nothing in the code seems to allow it to do so.

**Comment by formatBCE** [comment]
Ping on this. I have 3 UARTs configured, and whichever is written last, doesn't work.
Should i use GPIO1 and GPIO3 for third device, to share it with logger?

**Comment by ssieb** [comment]
You can't share it, you would have to disable serial logging.  All ESP32 variants have 3 hardware uarts, but esphome will currently only use uart0 if you specify GPIO1 and 3 for the pins.

**Comment by formatBCE** [comment]
> You can't share it, you would have to disable serial logging.  All ESP32 variants have 3 hardware uarts, but esphome will currently only use uart0 if you specify GPIO1 and 3 for the pins.

Yes, already realized it. Still, no software UART despite having it in docs...

**Comment by ShunichiSan** [comment]
I think they just need to update the docs since there's no intention at this point to add software UART.

**Comment by ssieb** [comment]
The note mentions it only for 8266, but I suppose the "Hardware UARTs" section could clarify that the software option only applies to 8266.

**Comment by Fabian-Schmidt** [comment]
FYI: Victron has documented their Bluetooth advertising protocol so you could switch from UART to Bluetooth for Victron devices. This will not solve the Software UART request, but  I am happy with a BLE link and therefore need two UART less on my ESP.

<https://github.com/Fabian-Schmidt/esphome-victron_ble>

**Comment by nagyrobi** [comment]
> The note mentions it only for 8266, but I suppose the "Hardware UARTs" section could clarify that the software option only applies to 8266.

Done: https://github.com/esphome/esphome-docs/pull/2734

**Comment by jhollowe** [comment]
As a note: if you set the logger baud rate to 0, it will not use up a hardware UART.

e.g.
```yaml
logger:
  level: DEBUG
  # disable putting logging on the HW UART
  baud_rate: 0
```

**Comment by vekexasia** [comment]
@jhollowe are you able to see the logs in wifi with logger level debug and baud_rate 0?

**Comment by ssieb** [comment]
@vekexasia yes, it only stops the logger sending data on the uart.

**Comment by JoHo94** [comment]
+1 for this feature. I want to use one esp to control multiple UARTS. So it would be awesome to use the 3 Hardware UARTs + some Software ones.

**Comment by bzumik1** [comment]
I would also love to see this feature implemented! 🙂

**Comment by zechnkaas** [comment]
+1 here (also like to add 4 Victron Devices and one DALY BMS)

did anyone get in the meantime try on of those to expand UART

[https://esphome.io/components/weikai.html](url)

**Comment by psolyca** [comment]
Nothing new on this topic !!

I'd like to use 3 UARTs on my ESP32

**Comment by nagyrobi** [comment]
Since then, ESP32-P4 was released. It should [support](https://docs.espressif.com/projects/esp-idf/en/stable/esp32p4/api-reference/peripherals/uart.html) 5 high performce uarts and 1 low power one.
ESP32-S2, ESP32-S3, ESP32-P4 also support [USB host interface](https://esphome.io/components/usb_host/). That means, you can connect [USB serial adapters](https://esphome.io/components/usb_uart/) to these, in addition of the hardware uarts. Even with multiple ports (up to 3 on ESP32-S2, ESP32-S3, making a total of 6 on these, or even more on P4)

**Comment by ssieb** [comment]
> I'd like to use 3 UARTs on my ESP32

Which variant?  Most support 3 uarts.

You can also add an external uart expander like https://esphome.io/components/weikai/

**Comment by psolyca** [comment]
I have the "classic" ESP32, a clone of WEMOS D1 mini ESP32.
There are 3 HW UARTs but only 2 are really accessible cause UART1 (GPIO9 - GPIO10) is used for flashmemory.
I could use other GPIO as softwareserial cause devices connected to my ESP32 are low baudrate (1000).
I do not know why buying an expension board when the device is capable of doing it... I do not want to hurt just asking.
If I can't, i will try to use an ESP8266 as UART expansion board, I just need a way to communicate between them.

Thanks anyway for this great tool.

**Comment by ssieb** [comment]
There are no fixed pins for the uarts.  I don't know where you got 9 and 10 from.  As long as you've disabled the serial logging, you can use all 3 uarts on almost any pins you want.

**Comment by francistheodorecatte** [comment]
Also requesting extending software UART support to the ESP32's. The ESP32C3 only has two hardware UARTs, and a software UART would negate needing extra hardware like a WeiKai IIC/UART bridge (which has caveats beyond being extra hardware) for a third UART.

A real use-case for this is (re)installing a Senseair S8 CO2 sensor in the Airgradient O-1PPT, where both hardware UARTs are already in use by the two PMS5003T particulate matter sensors.
